Docker common consolidation

DBendit opened the pull request:
##### SUMMARY
Fixes #49563 and #49564

Consolidates Go Boolean cleaning and minimum version checks into docker_common.py.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
docker_common
docker_network
docker_prune
docker_container
docker_swarm

##### ADDITIONAL INFORMATION
I can't get the docker_container integ tests passing locally (auto-remove fails for some reason), but docker_network's are running fine on my machine. Additionally, testing with old versions of docker-py resulted in seeing the expected errors coming out of the validation for both docker_network and docker_container.

Edit: docker_network and docker_swarm tests are passing. Manual testing for outdated versions is working as expected. I explicitly tested docker_container with an old docker-py and an `ipv4_address` network and it responded correctly, so the latest change to pass in the function for `detect_usage` is working as expected.

DBendit committed:
[docker] Consolidating Python Boolean conversion for Docker API (#49563)

DBendit committed:
[docker] Consolidating docker option min version checks (#49564)

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -952,6 +952,16 @@ def split_colon_ipv6(input, module):
     return result
 
 
+def detect_ipvX_address_usage():
```
It's better to define this one inline shortly before `option_minimal_versions`, so you have access to `client.module.params`.

felixfontein commented on lib/ansible/module_utils/docker_common.py:
```diff
@@ -741,3 +797,23 @@ def get_legacy_docker_diffs(self):
         return result
+
+
+def clean_booleans_for_docker_api(data):
```
How about including `dict` there somehow? Like `clean_dict_booleans_for_docker_api`?

DBendit replied:
"There are only two hard things in Computer Science: cache invalidation and naming things."

felixfontein commented:
@DBendit now that #49691 has been merged, there's one more module to adjust. (I wanted #49691 to be merged first because I really want to backport that to 2.7.)

DBendit committed:
[docker] Moving option min version checks out of docker_swarm (#49564)

Also renaming Boolean cleanup function and fixing docker_container minimum
version check for network interfaces.

DBendit commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -2856,8 +2764,20 @@ def detect_ipvX_address_usage():
+        if 'option_minimal_versions' in kwargs:
+            if 'ipvX_address_supported' in kwargs['option_minimal_versions']:
+                if 'detect_usage' not in kwargs['option_minimal_versions']['ipvX_address_supported']:
+                    kwargs['option_minimal_versions']['ipvX_address_supported']['detect_usage'] = detect_ipvX_address_usage
```
I'm not a huge fan of this, but I can't think of a better way to get this function passed in to the superclass. Any ideas?

felixfontein replied:
Hmmm, this is indeed somewhat ugly.

I would simply put it into `main()`. Or at least get rid of the `if`s and simply write
```
kwargs['option_minimal_versions']['ipvX_address_supported'] = dict(docker_py_version='1.9.0', usage_msg='ipv4_address or ipv6_address in networks', detect_usage=detect_ipvX_address_usage)
```
After all, nobody else is using `AnsibleDockerClientContainer`.

Alternatively, we could define `option_minimal_versions` here and pass it to the super-constructor. This will blow up the constructor a bit, but at least it will be a bit clearer :)

DBendit commented:
ready_for_review

felixfontein commented on lib/ansible/module_utils/docker_common.py:
```diff
@@ -416,6 +420,58 @@ def _handle_ssl_error(self, error):
 
+    def _get_minimal_versions(self, option_minimal_versions, ignore_params=None):
+        self.option_minimal_versions = dict()
+        for option, data in self.module.argument_spec.items():
```
Maybe I should have used
```suggestion
        for option in self.module.argument_spec:
```
:-)

DBendit committed:
Cleanup from PR feedback

felixfontein reviewed (approved):
LGTM shipit

felixfontein commented:
@DBendit thanks for this PR!
@gundalow thanks for merging!
